go/store/prolly/node_store.go

```diff
@@ -33,36 +41,235 @@ type NodeStore interface {
 
 	// Pool returns a buffer pool.
 	Pool() pool.BuffPool
+
+	// Format returns the types.NomsBinFormat of this NodeStore.
+	Format() *types.NomsBinFormat
 }
 
 type nodeStore struct {
 	store chunks.ChunkStore
+	cache *chunkCache
 	bp    pool.BuffPool
 }
 
 var _ NodeStore = nodeStore{}
 
+var sharedCache = newChunkCache(cacheSize)
+
 var sharedPool = pool.NewBuffPool()
 
 // NewNodeStore makes a new NodeStore.
 func NewNodeStore(cs chunks.ChunkStore) NodeStore {
-	return nodeStore{store: cs, bp: sharedPool}
+	return nodeStore{
+		store: cs,
+		cache: sharedCache,
+		bp:    sharedPool,
+	}
 }
 
 // Read implements NodeStore.
 func (ns nodeStore) Read(ctx context.Context, ref hash.Hash) (Node, error) {
+	c, ok := ns.cache.get(ref)
+	if ok {
+		return c.Data(), nil
+	}
+
 	c, err := ns.store.Get(ctx, ref)
+	if err != nil {
+		return nil, err
+	}
+	ns.cache.insert(c)
 	return c.Data(), err
 }
 
 // Write implements NodeStore.
 func (ns nodeStore) Write(ctx context.Context, nd Node) (hash.Hash, error) {
 	c := chunks.NewChunk(nd)
-	err := ns.store.Put(ctx, c)
-	return c.Hash(), err
+	if err := ns.store.Put(ctx, c); err != nil {
+		return hash.Hash{}, err
+	}
+	ns.cache.insert(c)
+	return c.Hash(), nil
 }
 
 // Pool implements NodeStore.
 func (ns nodeStore) Pool() pool.BuffPool {
 	return ns.bp
 }
+
+func (ns nodeStore) Format() *types.NomsBinFormat {
+	// todo(andy): read from |ns.store|
+	return types.Format_DOLT_1
+}
+
+type centry struct {
+	c    chunks.Chunk
+	i    int
```

Comment:
?

Reply:
"cache entry" stole this from Aaron's implementation in `dolthubapi`